src/printer.js

```diff
@@ -152,24 +129,134 @@ function isLastStatement(path) {
+  const nodeIndex = getNodeIndex(path);
+  return parentNodeBody && nodeIndex === parentNodeBody.length - 1;
+}
+
+function isPrevNodeInline(path) {
+  const nodeIndex = getNodeIndex(path);
+  const parentNodeBody = getParentNodeListProperty(path);
+  const prevNode = nodeIndex !== -1 && parentNodeBody[nodeIndex - 1];
+  return prevNode && prevNode.kind === "inline";
+}
+
+function isNextNodeInline(path) {
+  const parentNodeBody = getParentNodeListProperty(path);
+  const nodeIndex = getNodeIndex(path);
+  const nextNode = nodeIndex !== -1 && parentNodeBody[nodeIndex + 1];
+  return nextNode && nextNode.kind === "inline";
+}
+
+function lineShouldEndWithHardline(path) {
+  const node = path.getValue();
+  const nodeIndex = getNodeIndex(path);
+  return (
+    nodeIndex !== -1 &&
+    !isLastStatement(path) &&
+    node.kind !== "case" &&
+    !isNextNodeInline(path)
+  );
+}
+
+function lineShouldHaveStartPHPTag(path) {
+  return isFirstNodeInParentProgramNode(path) || isPrevNodeInline(path);
+}
+
+function lineShouldHaveEndPHPTag(path) {
+  return isNextNodeInline(path);
+}
+
+function fileShouldEndWithHardline(path) {
+  const node = path.getValue();
+  const isProgramNode = node.kind === "program";
+  return (
+    isProgramNode && node.children[node.children.length - 1].kind !== "inline"
+  );
+}
+
+function genericPrint(path, options, print) {
+  const node = path.getValue();
+  if (!node) {
+    return "";
+  } else if (typeof node === "string") {
+    return node;
+  }
+  const isProgramNode = node.kind === "program";
+  if (isProgramNode && node.children.length === 0) {
+    return concat(["<?php", hardline]);
+  }
+  const printed = printNode(path, options, print);
+  if (node.kind === "inline") {
+    return concat([printed]);
+  }
+  if (node.kind === "block") {
+    const nodeBody = getNodeListProperty(node);
+    if (nodeBody.length !== 0) {
+      return concat([
+        nodeBody[0].kind === "inline" ? "?>" : "",
+        printed,
+        nodeBody[nodeBody.length - 1].kind === "inline" ? "<?php " : ""
+      ]);
     }
-    return concat(parts);
-  });
-  return concat(printed);
+  }
+  return concat([
+    lineShouldHaveStartPHPTag(path)
+      ? concat([
+          "<?php",
+          isFirstNodeInParentProgramNode(path) ||
+          isLastNodeInParentNode(path) ||
+          !isNextNodeInline(path)
+            ? hardline
+            : " "
+        ])
+      : "",
+    printed,
+    lineShouldEndWithSemicolon(path) ? ";" : "",
+    lineShouldEndWithHardline(path)
+      ? concat([
+          sharedUtil.isNextLineEmpty(options.originalText, node, options)
+            ? hardline
+            : "",
+          hardline
+        ])
+      : "",
+    lineShouldHaveEndPHPTag(path)
+      ? concat([
+          isFirstNodeInParentNode(path) || !isPrevNodeInline(path)
+            ? hardline
+            : " ",
+          "?>"
+        ])
+      : "",
+    fileShouldEndWithHardline(path) ? hardline : ""
+  ]);
```

Comment:
Should be easy to read and maintenance :+1:

Reply:
agreed looks great 🎉